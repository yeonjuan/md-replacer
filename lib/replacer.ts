import Pipe = require("./pipe");
import createIndent = require("./create-indent");

interface Context {
  indent: ReturnType<typeof createIndent>;
}

type ReplaceFunc = (ctx: Context) => string;

class Replacer {
  static START_INDENT_REGEX = /([ \t]*)<!--\s*replace-start:[\s\S]*?-->.*/;
  static END_INDENT_REGEX = /([ \t]*)<!--\s*replace-end:[\s\S]*?-->.*/;
  static INNER_CONTENT_REGEX =
    /[ \t]*?<!--\s*replace-start:([\s\S]*?)-->([\s\S]*?)<!--\s*replace-end:([\s\S]*?)-->/g;

  private pipe = new Pipe<(input: string) => string>();
  constructor(private text: string) {}

  public replace(name: string, replaceFunc: ReplaceFunc): this {
    this.pipe.push((input) =>
      this.replaceCommentPart(input, name, replaceFunc)
    );
    return this;
  }

  public build(): string {
    const pipeline = this.pipe.pipeline();
    return pipeline(this.text);
  }

  private createContext(startCommentLine: string): Context {
    const indentStr = startCommentLine.replace(
      Replacer.START_INDENT_REGEX,
      "$1"
    );
    return {
      indent: createIndent(indentStr),
    };
  }

  private replaceCommentPart(
    text: string,
    name: string,
    replaceFunc: ReplaceFunc
  ): string {
    let replaced = text;
    const allMatched = text.matchAll(Replacer.INNER_CONTENT_REGEX);
    for (let [matched, startName, innerContent, endName] of allMatched) {
      let matchedName = null;
      const startTrimmed = startName.trim();
      const endTrimmed = endName.trim();
      if (startTrimmed === endTrimmed && startTrimmed === name) {
        matchedName = name;
      }
      if (matchedName) {
        const context = this.createContext(matched);
        const endIndentStr = matched.replace(Replacer.END_INDENT_REGEX, "$1");
        const isInline = innerContent.indexOf("\n") === -1;
        const newLineOrEmpty = isInline ? "" : "\n";
        replaced = replaced.replace(
          matched,
          matched.replace(
            innerContent,
            `${newLineOrEmpty}${replaceFunc(context)}${newLineOrEmpty}`
          )
        );
      }
    }
    return replaced;
  }
}

export = function create(text: string): Replacer {
  return new Replacer(text);
};
// 아이디어
// 파싱을 먼저 한다
/**
 * {
 *   name: string
 *   startComment: {
 *     start: number,
 *     end: number,
 *   },
 *   endComment {
 *     start: number,
 *     end: number
 *   },
 *   lines: string[]
 * }
 */
