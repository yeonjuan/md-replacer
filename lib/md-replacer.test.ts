import replacer = require("./md-replacer");

describe("md-replacer", () => {
  it("Inline", () => {
    const INPUT = "<!-- start: foo -->aaa <!-- end: foo -->";
    const result = replacer()
      .content(INPUT)
      .replace("foo", () => "foo")
      .build();
    expect(result).toMatchSnapshot();
  });

  it("Multiline", () => {
    const INPUT = `
<!-- start: foo -->
aaa
<!-- end: foo -->`;
    const result = replacer()
      .content(INPUT)
      .replace("foo", () => "foo")
      .build();
    expect(result).toMatchSnapshot();
  });

  it("Mixed", () => {
    const INPUT = `
before
<!-- start: foo -->
aaa
<!-- end: foo -->between<!-- start: bar -->aaa<!-- end: bar -->after`;
    const result = replacer()
      .content(INPUT)
      .replace("foo", () => "foo")
      .replace("bar", () => "bar")
      .build();
    expect(result).toMatchSnapshot();
  });

  it("replace", () => {
    const INPUT = `
before
  <!-- start: foo -->
  aaa
  <!-- end: foo -->
ebd`;
    const result = replacer()
      .content(INPUT)
      .replace("foo", ({ indent }) => indent("foo"))
      .build();
    expect(result).toMatchSnapshot();
  });
});
