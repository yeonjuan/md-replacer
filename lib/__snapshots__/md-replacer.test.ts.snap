// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`md-replacer Inline 1`] = `"<!-- start: foo -->foo<!-- end: foo -->"`;

exports[`md-replacer Mixed 1`] = `
"
before
<!-- start: foo -->
foo
<!-- end: foo -->between<!-- start: bar -->bar<!-- end: bar -->after"
`;

exports[`md-replacer Multiline 1`] = `
"
<!-- start: foo -->
foo
<!-- end: foo -->"
`;

exports[`md-replacer replace 1`] = `
"
before
  <!-- start: foo -->
  foo
  <!-- end: foo -->
ebd"
`;
